my-megaparsec-project/test/OptimizerPropertySpec.hs

```diff
@@ -0,0 +1,395 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module OptimizerPropertySpec (spec) where
+
+import Test.Hspec
+import Test.QuickCheck
+import Test.QuickCheck.Monadic
+import qualified Data.Map as M
+import qualified Data.Set as S
+import Data.Text (Text, pack)
+import Control.Monad.State
+import Data.List (nub)
+import Debug.Trace
+import Control.Exception (evaluate, try, SomeException, fromException, ArithException)
+import Test.QuickCheck.Monadic (monadicIO, run, assert)
+import Data.Typeable (typeOf)
+
+import Expression
+import Statement
+import Optimizer
+
+-- Set this to True to enable more verbose logging
+verboseLogging :: Bool
+verboseLogging = False
+
+-- Helper for conditional logging
+logTrace :: String -> a -> a
+logTrace msg = if verboseLogging then trace msg else id
+
+-- | Generate arbitrary variable names
+genVarName :: Gen Text
+genVarName = do
+  c <- elements ['a'..'z']
+  return $ pack [c]
+
+-- | Generate arbitrary integer literals within a reasonable range
+genIntLit :: Gen Int
+genIntLit = choose (-100, 100)
+
+-- Helper for checking if expression is zero
+isZero :: Expr -> Bool
+isZero (IntLit 0) = True
+isZero _ = False
+
+-- | Generate arbitrary expressions
+instance Arbitrary Expr where
+  arbitrary = sized genExpr
+  
+  -- Shrink implementation to simplify failing test cases
+  shrink (Sum a b) = [a, b] ++ [Sum a' b | a' <- shrink a] ++ [Sum a b' | b' <- shrink b]
+  shrink (Sub a b) = [a, b] ++ [Sub a' b | a' <- shrink a] ++ [Sub a b' | b' <- shrink b]
+  shrink (Prod a b) = [a, b] ++ [Prod a' b | a' <- shrink a] ++ [Prod a b' | b' <- shrink b]
+  shrink (Div a b) = [a] ++ [Div a' b | a' <- shrink a] ++ [Div a b' | b' <- shrink b, not $ isZero b']
+  shrink (Neg a) = [a] ++ [Neg a' | a' <- shrink a]
+  shrink _ = []
```

Comment:
I wonder if this is necessary. I feel that it's easier to filter out division by zero when you use the generator. Especially given that an expression with `/ (1-1)` can still be generated, and this won't understand that it has.